contracts/contract/src/main.rs

```diff
@@ -0,0 +1,601 @@
+    if let Some(minter_list) = minter_list {
+        for account_key in minter_list {
+            badge_map.insert(account_key, SecurityBadge::Minter);
+        }
+    }
+    if let Some(admin_list) = admin_list {
+        for account_key in admin_list {
+            badge_map.insert(account_key, SecurityBadge::Admin);
+        }
+    }
+    if let Some(none_list) = none_list {
+        for account_key in none_list {
+            badge_map.insert(account_key, SecurityBadge::None);
+        }
+    }
+
+    let caller = get_immediate_caller();
+    badge_map.remove(&caller);
+
+    change_sec_badge(&badge_map);
+    events::record_event_dictionary(Event::ChangeSecurity(ChangeSecurity {
+        admin: caller,
+        sec_change_map: badge_map,
+    }));
+}
+
+#[no_mangle]
+fn change_events_mode() {
+    sec_check(vec![SecurityBadge::Admin]);
+    let events_mode: EventsMode = EventsMode::try_from(get_named_arg::<u8>(ARG_EVENTS_MODE))
+        .unwrap_or_revert_with(Cep18Error::InvalidEventsMode);
+    let old_events_mode: EventsMode = EventsMode::try_from(get_stored_value::<u8>(ARG_EVENTS_MODE))
+        .unwrap_or_revert_with(Cep18Error::InvalidEventsMode);
+    if events_mode == old_events_mode {
+        return;
+    }
+    let events_mode_u8 = events_mode as u8;
+    put_key(ARG_EVENTS_MODE, storage::new_uref(events_mode_u8).into());
+    init_events();
+
+    events::record_event_dictionary(Event::ChangeEventsMode(ChangeEventsMode {
+        events_mode: events_mode_u8,
+    }));
+}
+
+pub fn upgrade(name: &str) {
+    let entry_points = generate_entry_points();
+
+    let package_key_name = &format!("{PREFIX_CEP18}_{PREFIX_CONTRACT_PACKAGE_NAME}_{name}");
+    let contract_key_name = &format!("{PREFIX_CEP18}_{PREFIX_CONTRACT_NAME}_{name}");
+
+    let old_contract_package_hash = match runtime::get_key(package_key_name)
+        .unwrap_or_revert_with(Cep18Error::FailedToGetOldPackageKey)
+    {
+        Key::Hash(contract_hash) => contract_hash,
+        Key::AddressableEntity(EntityAddr::SmartContract(contract_hash)) => contract_hash,
+        Key::SmartContract(package_hash) => package_hash,
+        _ => revert(Cep18Error::MissingPackageHashForUpgrade),
+    };
+    let contract_package_hash = PackageHash::new(old_contract_package_hash);
+
+    let previous_contract_hash = match runtime::get_key(contract_key_name)
+        .unwrap_or_revert_with(Cep18Error::FailedToGetOldContractHashKey)
+    {
+        Key::Hash(contract_hash) => contract_hash,
+        Key::AddressableEntity(EntityAddr::SmartContract(contract_hash)) => contract_hash,
+        _ => revert(Cep18Error::MissingContractHashForUpgrade),
+    };
+    let converted_previous_contract_hash = AddressableEntityHash::new(previous_contract_hash);
+
+    let events_mode = get_optional_named_arg_with_user_errors::<u8>(
+        ARG_EVENTS_MODE,
+        Cep18Error::InvalidEventsMode,
+    );
+
+    let version_value_uref = get_uref_with_user_errors(
```

Comment:
is it ok to fail here? can a CEP-18 contract from protocol 1.x be upgraded in this case?

Reply:
yes any version has PREFIX_CONTRACT_VERSION, it should fail if upgrading a contract that has no version in storage which should never ever be the case, alle versions ahve a version key, be it u32 for 1.x and String for 2.0

the is a trick below to cast to a string and into a u32 the uref

`// If stored version is a non empty string (and not a u32), it means it is already a 2.0
    // version, do not add message topics then, as already set when installed``